### PR TITLE
Feature/env application

### PR DESCRIPTION
**Pull request by manas-droid**
Story ID: **38429**

**1. Changed the Add Application Pop in the Environment page to a pop up of check boxes
2. Added a functionality to publish applications to other environments**

Thank you for your contribution.
Before submitting this PR, please make sure:

- [ ] PR description and commit message should describe the changes done in this PR
- [ ] Verify the PR is pointing to correct branch i.e. Release or Beta branch if the code fix is for specific release , else point it to master
- [ ] Latest Code from master or specific release branch is merged to your branch
- [ ] No unwanted\commented\junk code is included
- [ ] No new warning upon build solution
- [ ] Code Summary\Comments are added to my code which explains what my code is doing
- [ ] Existing unit test cases are passed
- [ ] New Unit tests are added for your development
- [ ] Sanity Tests are successfully executed for New and Existing Functionality
- [ ] Verify that changes are compatible with all relevant browsers and platforms.
- [ ] After creating pull request there should not be any conflicts
- [ ] Resolve all Codacy comments
- [ ] Builds and checks are passed before PR is sent for review
- [ ] Resolve code review comments
- [ ] Update the Help Library document to match any feature changes

## Summary by CodeRabbit

- **New Features**
	- Added new user messages for application publishing and platform availability.
	- Enhanced application listing within environments with interactive display features.

- **Enhancements**
	- Updated the property setter for item image types in repository items, improving customization.

- **Tests**
	- Improved testing framework with new tests for application platform addition and environment folder operations.
	- Adjusted existing tests for better functionality and clarity.

- **Refactor**
	- Modified window close button visibility logic for a more intuitive user interface.

### CODE REVIEW

**Review by Maheshkale447 — CHANGES_REQUESTED** [review]

File: Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentTreeItem.cs [hunk 1]
```diff
@@ -110,17 +117,55 @@ void ITreeViewItem.SetTools(ITreeView TV)
 
         private void AddApplication(object sender, RoutedEventArgs e)
         {
+            var ApplicationPlatforms = WorkSpace.Instance.Solution.ApplicationPlatforms.Where((app) => !ProjEnvironment.CheckIfApplicationPlatformExists(app.Guid , app.AppName))?.ToList();
+
+
+
+            if( ApplicationPlatforms == null || ApplicationPlatforms?.Count == 0)
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentTreeItem.cs:124]: ok will change this

File: Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentTreeItem.cs [hunk 2]
```diff
@@ -110,17 +117,55 @@ void ITreeViewItem.SetTools(ITreeView TV)
 
         private void AddApplication(object sender, RoutedEventArgs e)
         {
+            var ApplicationPlatforms = WorkSpace.Instance.Solution.ApplicationPlatforms.Where((app) => !ProjEnvironment.CheckIfApplicationPlatformExists(app.Guid , app.AppName))?.ToList();
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentTreeItem.cs:120]: I need ToList() to convert the List to observable list using the GingerCore.General.ConvertListToObservableList() function

File: Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentApplicationList.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        }
+        public void ShowAsWindow(eWindowShowStyle windowStyle = eWindowShowStyle.Dialog, bool ShowCancelButton = true)
+        {
+            Button okBtn = new Button();
+            okBtn.Content = "Ok";
+            ObservableList<Button> winButtons = new ObservableList<Button>();
+            winButtons.Add(okBtn);
+            WeakEventManager<ButtonBase, RoutedEventArgs>.AddHandler(source: okBtn, eventName: nameof(ButtonBase.Click), handler: OKButton_Click);
+
+            GingerCore.General.LoadGenericWindow(ref _pageGenericWin, App.MainWindow, windowStyle, this.Title, this, winButtons, ShowCancelButton, "Cancel");
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentApplicationList.xaml.cs:51]: The cancel button here is the upper button with the "cross" icon. There is another close button which is set when you want a "close" button next to the "ok".

File: Ginger/Ginger/Environments/AppsListPage.xaml.cs [hunk 4]
```diff
@@ -104,6 +106,7 @@ private void SetGridView()
 
         private void SetGridData()
         {
+            AppEnvironmnet.Applications.ForEach((app) => app.SetPlatFormImage(WorkSpace.Instance.Solution.ApplicationPlatforms));
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/Ginger/Environments/AppsListPage.xaml.cs:109]: Hello , I think, I will just remove this function from the PR and handle it in the following PR : 
https://github.com/Ginger-Automation/Ginger/pull/3550/files#:~:text=public%20void-,SetDataFromAppPlatform,-(ObservableList%3C

As additional attributes are added there which will help to remove the setter in the ItemImageType.

File: Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentApplicationList.xaml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+﻿<Page x:Class="Ginger.SolutionWindows.TreeViewItems.EnvironmentsTreeItems.EnvironmentApplicationList"
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentApplicationList.xaml:1]: Should I add the licence in the Xaml or xaml.cs (as in all the our code only .xaml.cs has the licence).

File: Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentTreeItem.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentTreeItem.cs:124]: Done

File: Ginger/GingerCoreCommon/EnvironmentLib/EnvApplication.cs [hunk 6]
```diff
@@ -110,5 +119,19 @@ public override string ItemNameField
...
+            ApplicationPlatform applicationPlatform =  ApplicationPlatforms.FirstOrDefault((app)=>app.Guid.Equals(this.ParentGuid));
+            if(applicationPlatform != null)
+            {
+                this.ItemImageType = applicationPlatform.PlatformImage;
+                this.Name = applicationPlatform.AppName;
+            }
+            else
+            {
+                this.ItemImageType = eImageType.Application;
+            }
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerCoreCommon/EnvironmentLib/EnvApplication.cs:134]: A case for when applicationPlatform equals to null is already handled and is not actually considered an error/exception.

File: Ginger/GingerTest/EnvironemntsLib/EnvsTest.cs [hunk 7]
```diff
@@ -130,6 +132,31 @@ public void AddEnvToFileSystemWillShowinEnvsTree()
...
+            ObservableList<EnvApplication> expectedApplicationList = e1.Applications;
+
+            string txt = e1.RepositorySerializer.SerializeToString(e1);
+            string fileName = Path.Combine(SolutionFolder, @"Environments\aaa.Ginger.Environment.xml");
+            File.WriteAllText(fileName, txt);
+
+            ProjEnvironment aaa = EnvsPOM.SelectEnvironment("aaa");
+
+            Assert.AreEqual(expectedApplicationList.Count, aaa.Applications.Count);
+        }
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerTest/EnvironemntsLib/EnvsTest.cs:159]: The test cases in this file are not run

File: Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentApplicationList.xaml.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentApplicationList.xaml.cs]: ok , I have implemented this

File: Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentApplicationList.xaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (Maheshkale447) [reply Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentApplicationList.xaml:1]: only in xaml.cs

File: Ginger/GingerCore/GeneralLib/GenericWindow.xaml.cs [hunk 8]
```diff
@@ -219,9 +219,9 @@ public partial class GenericWindow : Window
 
             //close buttons handling
             mCloseEventHandler = closeEventHandler;
-            if (!showClosebtn)
+            if (!dontShowCloseBtn)
```
Comment (Maheshkale447) [inline Ginger/GingerCore/GeneralLib/GenericWindow.xaml.cs:222]: Why are we changing it to be visible when its visible by default, also please check other places wherever this window is being used

File: Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentApplicationList.xaml.cs [hunk 9]
```diff
@@ -40,20 +57,11 @@ private void SetGridView()
...
+        public void ShowAsWindow(eWindowShowStyle windowStyle = eWindowShowStyle.Dialog, bool DontShowCloseBtn = false)
         {
-            Button okBtn = new Button();
-            okBtn.Content = "Ok";
-            ObservableList<Button> winButtons = new ObservableList<Button>();
-            winButtons.Add(okBtn);
-            WeakEventManager<ButtonBase, RoutedEventArgs>.AddHandler(source: okBtn, eventName: nameof(ButtonBase.Click), handler: OKButton_Click);
 
-            GingerCore.General.LoadGenericWindow(ref _pageGenericWin, App.MainWindow, windowStyle, this.Title, this, winButtons, ShowCancelButton, "Cancel");
+            GingerCore.General.LoadGenericWindow(ref _pageGenericWin, App.MainWindow, windowStyle, this.Title, this, null, DontShowCloseBtn, "ok");
```
Comment (Maheshkale447) [inline Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentApplicationList.xaml.cs:63]: change ok label to Ok or Done

**Review by Maheshkale447 — CHANGES_REQUESTED** [review]

File: Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentApplicationList.xaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentApplicationList.xaml:1]: Ok. I have already added

File: Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentApplicationList.xaml.cs (hunk 9 above)
Comment (manas-droid) [reply Ginger/Ginger/SolutionWindows/TreeViewItems/EnvironmentsTreeItems/EnvironmentApplicationList.xaml.cs:63]: Ok

File: Ginger/GingerCore/GeneralLib/GenericWindow.xaml.cs (hunk 8 above)
Comment (manas-droid) [reply Ginger/GingerCore/GeneralLib/GenericWindow.xaml.cs:222]: I checked the xaml and by default it is Collapsed. I checked other places and every 
"true" or no attribute is passed for "showCloseBtn"

**Review by Maheshkale447 — APPROVED** [review]

**Pull request merged**